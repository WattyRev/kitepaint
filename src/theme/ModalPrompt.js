import React from "react";
import PropTypes from "prop-types";
import styled, { keyframes } from "styled-components";
import BodyPortal from "./utilities/BodyPortal";
import Label from "./Label";
import Input from "./Input";
import Button from "./Button";

const modalAnimation = keyframes`
  0% {
    opacity: 0
  }

  15% {
    opacity: 0
  }

  100% {
    opacity: 1
  }
`;

const StyledModal = styled.div`
  position: fixed;
  top: 50px;
  left: 50%;
  transform: translateX(-50%);
  max-width: 300px;
  background: ${props => props.theme.colors.white};
  z-index: 100;
  padding: 16px;
  border: 1px solid ${props => props.theme.colors.gray};
  border-radius: 4px;
  animation: ${modalAnimation} 0.8s;

  .buttons {
    display: flex;

    button {
      flex-grow: 1;
      margin: 8px 8px 0;
    }
  }
`;

const backdropAnimation = keyframes`
  0% {
    background-position: 0% 0%;
  },
  100% {
    background-position: 100% 100%;
  }
`;

const ModalBackdrop = styled.div`
  position: fixed;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  background: linear-gradient(-45deg, rgba(0, 0, 0, 0.7), rgba(0, 0, 0, 0));
  background-size: 400% 400%;
  background-position: 100% 100%;
  z-index: 100;
  animation: ${backdropAnimation} 0.6s;
`;

class ModalPrompt extends React.Component {
  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    onCancel: PropTypes.func,
    modalMessage: PropTypes.string.isRequired,
    submitText: PropTypes.string,
    cancelText: PropTypes.string,
    children: PropTypes.func.isRequired
  };

  static defaultProps = {
    submitText: "Submit",
    cancelText: "Cancel",
    onCancel: () => {}
  };

  state = {
    isOpen: false,
    value: ""
  };

  handleSubmit = event => {
    event.preventDefault();
    if (!this.state.value) {
      return;
    }
    this.props.onSubmit(this.state.value);
    this.setState({ isOpen: false, value: "" });
  };

  handleCancel = () => {
    this.setState({ isOpen: false, value: "" });
    this.props.onCancel();
  };

  handleValueChange = event => {
    this.setState({ value: event.target.value });
  };

  render() {
    const data = {
      actions: {
        open: () => this.setState({ isOpen: true })
      },
      props: {
        isOpen: this.state.isOpen
      }
    };
    return (
      <React.Fragment>
        {this.props.children(data)}
        {this.state.isOpen && (
          <BodyPortal>
            <ModalBackdrop onClick={this.handleCancel} />
            <StyledModal>
              <form onSubmit={this.handleSubmit}>
                <Label>{this.props.modalMessage}</Label>
                <Input
                  value={this.state.value}
                  onChange={this.handleValueChange}
                />
                <div className="buttons">
                  <Button type="submit" isPrimary>
                    {this.props.submitText}
                  </Button>{" "}
                  <Button type="button" onClick={this.handleCancel}>
                    {this.props.cancelText}
                  </Button>
                </div>
              </form>
            </StyledModal>
          </BodyPortal>
        )}
      </React.Fragment>
    );
  }
}

export default ModalPrompt;
