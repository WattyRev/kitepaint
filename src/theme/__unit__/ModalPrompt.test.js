import React from "react";
import { mount, shallow } from "enzyme";
import Theme from "../../theme";
import ModalPrompt, { StyledModal } from "../ModalPrompt";

describe("ModelPrompt", () => {
  describe("StyledModal", () => {
    it("renders", () => {
      expect.assertions(1);
      const wrapper = mount(<StyledModal theme={Theme} />);
      expect(wrapper.find("div")).toHaveLength(1);
    });
  });

  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      onSubmit: jest.fn(),
      message: "Do the thing."
    };
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(
      <ModalPrompt {...defaultProps}>
        {modal => <div className="target" onClick={modal.actions.open} />}
      </ModalPrompt>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("does not display the modal by default", () => {
    expect.assertions(1);
    const wrapper = shallow(
      <ModalPrompt {...defaultProps}>
        {modal => <div className="target" onClick={modal.actions.open} />}
      </ModalPrompt>
    );
    expect(wrapper.find(".testing_modal")).toHaveLength(0);
  });
  it("opens a modal when the target is clicked", () => {
    expect.assertions(1);
    const wrapper = shallow(
      <ModalPrompt {...defaultProps}>
        {modal => <div className="target" onClick={modal.actions.open} />}
      </ModalPrompt>
    );
    wrapper.find(".target").simulate("click");
    expect(wrapper.find(".testing_modal")).toHaveLength(1);
  });
  it("Closes the modal when clicking on the cancel button", () => {
    expect.assertions(1);
    const wrapper = shallow(
      <ModalPrompt {...defaultProps}>
        {modal => <div className="target" onClick={modal.actions.open} />}
      </ModalPrompt>
    );
    wrapper.find(".target").simulate("click");
    wrapper.find(".testing_cancel").simulate("click");
    expect(wrapper.find(".testing_modal")).toHaveLength(0);
  });
  it("Closes the modal when clicking on the backdrop", () => {
    expect.assertions(1);
    const wrapper = shallow(
      <ModalPrompt {...defaultProps}>
        {modal => <div className="target" onClick={modal.actions.open} />}
      </ModalPrompt>
    );
    wrapper.find(".target").simulate("click");
    wrapper.find(".testing_backdrop").simulate("click");
    expect(wrapper.find(".testing_modal")).toHaveLength(0);
  });
  it("submits with the value and closes the modal", () => {
    expect.assertions(3);
    const wrapper = shallow(
      <ModalPrompt {...defaultProps}>
        {modal => <div className="target" onClick={modal.actions.open} />}
      </ModalPrompt>
    );
    wrapper.find(".target").simulate("click");
    wrapper.find(".testing_prompt-value").simulate("change", {
      target: {
        value: "boogers"
      }
    });
    wrapper.find(".testing_submit").simulate("submit", {
      preventDefault: jest.fn()
    });
    expect(defaultProps.onSubmit).toHaveBeenCalled();
    expect(defaultProps.onSubmit.mock.calls[0][0]).toEqual("boogers");
    expect(wrapper.find(".testing_modal")).toHaveLength(0);
  });
});
