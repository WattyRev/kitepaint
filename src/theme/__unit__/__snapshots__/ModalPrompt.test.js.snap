// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelPrompt renders 1`] = `
<Modal
  isOpen={false}
  modalContent={
    <ForwardRef
      className="testing_modal"
    >
      <form
        className="testing_submit"
        onSubmit={[Function]}
      >
        <ForwardRef>
          Do the thing.
        </ForwardRef>
        <ForwardRef
          className="testing_prompt-value"
          onChange={[Function]}
          value=""
        />
        <div
          className="buttons"
        >
          <Button
            isPrimary={true}
            type="submit"
          >
            Submit
          </Button>
           
          <Button
            className="testing_cancel"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </Button>
        </div>
      </form>
    </ForwardRef>
  }
  onBackdropClick={[Function]}
>
  <div
    className="target"
    onClick={[Function]}
  />
</Modal>
`;
